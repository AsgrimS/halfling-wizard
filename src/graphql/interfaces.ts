import {
  Class,
  Subclass,
  Proficiency,
  ClassProficiency_Choices,
  AbilityScoreSkills,
  ClassSpellcasting,
  ClassStarting_Equipment,
} from "./schema";

export interface HeroClassesData {
  classes: Class[];
}

export interface HeroClassData {
  class: Class;
}

export interface SubclassesData {
  subclasses: Subclass[];
}

export interface ProficienciesData {
  proficiencies: Proficiency[];
}

export interface ProficienciesChoicesData {
  proficienciesChoices: ClassProficiency_Choices[];
}
export interface SavingThrowsData {
  saving_throws: AbilityScoreSkills[];
}

export interface ClassSpellcastingData {
  spellcasting: ClassSpellcasting;
}

export interface StartingEquipmentData {
  startingEquipments: ClassStarting_Equipment[];
}
