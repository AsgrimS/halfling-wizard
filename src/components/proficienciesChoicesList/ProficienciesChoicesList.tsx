import React from "react";
import { v4 as uuidv4 } from "uuid";
import { ProficienciesChoicesData } from "../../graphql/interfaces";

const ProficienciesList: React.FC<ProficienciesChoicesData> = ({
  proficienciesChoices,
}) => {
  return (
    <div data-testid="proficiencies-choices-list">
      <span>Proficiencies Choices:</span>
      <br />
      {proficienciesChoices.map(({ from, choose }) => (
        <React.Fragment key={uuidv4()}>
          <span>Choose {choose}</span>
          <ul>
            {from!.map(({ name, index }) => (
              <li key={index}>{name}</li>
            ))}
          </ul>
        </React.Fragment>
      ))}
    </div>
  );
};

export default ProficienciesList;
