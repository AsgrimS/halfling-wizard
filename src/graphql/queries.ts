import { gql } from "@apollo/client";

export const HERO_CLASS = gql`
  query GetHeroesClasses($classIndex: String!) {
    class(filter: { name: $classIndex }) {
      hit_die
      index
      name
      proficiencies {
        index
        name
      }
      saving_throws {
        index
        name
      }
      proficiency_choices {
        choose
        from {
          name
          index
        }
      }
      spellcasting {
        level
        spellcasting_ability {
          index
          name
        }
      }
      starting_equipment {
        equipment {
          name
          index
        }
        quantity
      }
      starting_equipment_options {
        choose
        from {
          equipment {
            name
            index
          }
          quantity
        }
      }
      subclasses {
        index
        name
      }
    }
  }
`;

export const HEROES_CLASSES = gql`
  query GetHeroesClasses {
    classes {
      name
      index
    }
  }
`;
