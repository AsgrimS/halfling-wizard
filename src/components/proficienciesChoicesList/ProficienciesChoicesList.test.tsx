import React from "react";
import { render, screen, cleanup } from "@testing-library/react";

import ProficienciesChoicesList from "./ProficienciesChoicesList";
import { ClassProficiency_Choices } from "../../graphql/schema";

describe("ProficienciesChoicesList component", () => {
  afterEach(() => {
    cleanup();
  });

  it("should render a list of proficiencies choices", () => {
    const mockNames = [
      "Flying",
      "Invisibility",
      "Reading minds",
      "Super strength",
      "Dexterity like Batman",
    ];
    const proficienciesChoicesMockData: ClassProficiency_Choices[] = [
      {
        choose: 1,
        from: [
          { name: mockNames[0], index: mockNames[0].toLocaleLowerCase() },
          { name: mockNames[1], index: mockNames[1].toLocaleLowerCase() },
        ],
      },
      {
        choose: 2,
        from: [
          { name: mockNames[2], index: mockNames[2].toLowerCase() },
          { name: mockNames[3], index: mockNames[3].toLowerCase() },
          { name: mockNames[4], index: mockNames[4].toLowerCase() },
        ],
      },
    ];

    render(
      <ProficienciesChoicesList
        proficienciesChoices={proficienciesChoicesMockData}
      />
    );
    const proficienciesChoicesListElement = screen.getByTestId(
      "proficiencies-choices-list"
    );
    const liElements = proficienciesChoicesListElement.getElementsByTagName(
      "li"
    );
    const spanElements = proficienciesChoicesListElement.getElementsByTagName(
      "span"
    );
    const spanChooseElements = Array(spanElements).splice(1);

    expect(proficienciesChoicesListElement).toBeInTheDocument();
    expect(liElements.length).toEqual(5);

    for (let i = 0; i < spanChooseElements.length; i += 1) {
      expect(spanChooseElements[i]).toContainHTML(
        `<span>Choose ${proficienciesChoicesMockData[i].choose}</span>`
      );
    }

    for (let i = 0; i < liElements.length; i += 1) {
      expect(liElements[i]).toContainHTML(`<li>${mockNames[i]}</li>`);
    }

    expect(proficienciesChoicesListElement).toMatchSnapshot();
  });
});
