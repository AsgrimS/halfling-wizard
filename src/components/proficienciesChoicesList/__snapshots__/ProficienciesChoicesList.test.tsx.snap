// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProficienciesChoicesList component should render a list of proficiencies choices 1`] = `
<div
  data-testid="proficiencies-choices-list"
>
  <span>
    Proficiencies Choices:
  </span>
  <br />
  <span>
    Choose 
    1
  </span>
  <ul>
    <li>
      Flying
    </li>
    <li>
      Invisibility
    </li>
  </ul>
  <span>
    Choose 
    2
  </span>
  <ul>
    <li>
      Reading minds
    </li>
    <li>
      Super strength
    </li>
    <li>
      Dexterity like Batman
    </li>
  </ul>
</div>
`;
